### PR TITLE
GLOB-26190 - adding cache service to nodule-graphql

### PR DESCRIPTION
**Pull request by unorthodoxgeek**

### CODE REVIEW

File: src/services/caching/JSONStringify.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/* Stringify an object
+   Comatabile to the BE services swagger (Should return the exact same string)
+   based on flask with JSONIFY_PRETTYPRINT_REGULAR turned on
+   see: https://github.com/pallets/flask/blob/master/flask/json/__init__.py
+*/
+export default function JSONStringify(obj) {
+    return `${JSON.stringify(obj, null, 2).replace(/,/g, ', ')}\n`;
```
Comment (pelegdvir) [inline src/services/caching/JSONStringify.js]: Seems like magic here. Why do we use 2?

File: src/services/caching/__tests__/cached.test.js [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+                enabled: true,
+                ttl: 10,
+            },
+            logging: {
+                cache: {
+                    enabled: false,
+                },
+            },
+        },
+        logger: {
```
Comment (pelegdvir) [inline src/services/caching/__tests__/cached.test.js]: why do we need this?

File: src/services/caching/__tests__/spooky128.test.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import spooky128 from '../spooky128';
+
+describe('spooky', () => {
+    it('creates microcosm compatible cache', () => {
+        // microcosm has the same test
```
Comment (pelegdvir) [inline src/services/caching/__tests__/spooky128.test.js]: Link?

File: src/services/caching/__tests__/types.test.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
+/* Test caching config.
+ */
+import Enum from 'enum';
+import { CachingSpec, ANY_UUID } from '../types';
+
+const CachedObjectType = new Enum([
+    // lethe events (modulo snake case)
```
Comment (pelegdvir) [inline src/services/caching/__tests__/types.test.js]: hmmm

File: src/services/caching/__tests__/types.test.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    'userEvent',
+]);
+
+const CachingConfig = {
+    'ariza.userEvent.search': new CachingSpec({
```
Comment (pelegdvir) [inline src/services/caching/__tests__/types.test.js]: what is ariza?

File: src/services/caching/index.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+/* Configure caching.
+ */
+import { get } from 'lodash';
+import { bind, getContainer, getMetadata } from '@globality/nodule-config';
+
+export default function setCache() {
+    const { config, terminal, createCacheClient } = getContainer();
+    // fix-up caching configuration
```
Comment (pelegdvir) [inline src/services/caching/index.js]: ?

File: src/services/caching/operations.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
+/* Caching operations.
+ */
+
+
+/* Store a value at a key with a ttl.
+ *
+ * Uses memached `Add` operation, meaning that the styx will never overwrite
```
Comment (pelegdvir) [inline src/services/caching/operations.js]: This is a memached specific function, no?

File: src/services/caching/spooky128.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import Spooky from 'node-spookyhash-v2';
+
+// compatibility to python-spooky
```
Comment (pelegdvir) [inline src/services/caching/spooky128.js]: why do we need it?

File: src/services/caching/types.js [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+export const ANY_SINGLE_ITEM_LIST = Object();
+export const ANY_UUID = Object();
+
+
+/* Enumeration of cacheable object types.
+ *
+ * A subset of these type names are invalidable in `lethe`, so names should not be
+ * changed without coordination.
+ *
+ * See: https://github.com/globality-corp/lethe/blob/develop/lethe/models/cached_object_type.py
```
Comment (pelegdvir) [inline src/services/caching/types.js]: hmm

File: src/services/caching/wrapper.js [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        }
+        return [serviceData, CacheResult.noop];
+    } catch (err) {
+        const result = CacheResult.error;
+        logger.warning(req, err.message, {
+            error: err,
+            key,
+            args,
+        });
+        trace(spec, req, args, key, result);
```
Comment (pelegdvir) [inline src/services/caching/wrapper.js]: ?

File: src/services/caching/__tests__/types.test.js [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+
+describe('CachingSpec', () => {
+    const spec = CachingConfig['ariza.userEvent.search'];
+
+    it('creates keys', () => {
+        expect(
+            spec.createKey(),
+        ).toEqual(
+            '8a471b3a-6ae1-5e32-9693-cd641e9a8f09',
```
Comment (pelegdvir) [inline src/services/caching/__tests__/types.test.js]: can we separate the keys-tests from the rest of the tests?

File: src/services/caching/__tests__/JSONStringify.test.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        };
+        expect(
+            JSONStringify(obj),
+        ).toEqual(`{
+  "count": 2, 
+  "items": [
+    {
+      "_links": {
+        "child:prompt": {
+          "href": "http://dodona.dev.globality.io/api/v1/prompt_context?prompt_context_id=442160e0-74bb-482d-b229-c65a139a2092"
```
Comment (pelegdvir) [inline src/services/caching/__tests__/JSONStringify.test.js]: hmm

File: src/services/README.md [hunk 13]
```diff
@@ -83,6 +83,59 @@ to be included during batched requests, the `assignArgs` parameter can be used.
     }
     bind('serviceConfig.batch', batchConfig);
 
+### Cache
+
```
Comment (pelegdvir) [inline src/services/README.md]: Please expand:
Why do we use the cache?
What is `CachingSpec`? etc...

File: src/services/README.md [hunk 14]
```diff
@@ -83,6 +83,59 @@ to be included during batched requests, the `assignArgs` parameter can be used.
...
+    import Cache from 'memcached-promisify';
+    import { bind } from '@globality/nodule-config';
+
+    function createCacheClient(cacheHost, memcachedConfig) {
+        const client = new Cache(
+            { cacheHost },
+            memcachedConfig,
+        );
+        client.add = (key, data, lifetime) => cachePromise(
```
Comment (pelegdvir) [inline src/services/README.md]: `cachePromise?`

Comment (pelegdvir) [inline src/services/README.md]: Should we update `memcached-promisify`?

File: src/services/README.md [hunk 15]
```diff
@@ -83,6 +83,59 @@ to be included during batched requests, the `assignArgs` parameter can be used.
...
+
+    import { PAGINATION } from '../../constants';
+    import {
+        ANY_NOT_NULL,
+        ANY_PARAMETER,
+        ANY_UUID,
+        CachingSpec,
+    } from '@globality/nodule-graphql';
+
+    const CachedObjectType = new Enum([
```
Comment (pelegdvir) [inline src/services/README.md]: Do we have to use `CachedObjectType`? Why?

File: src/services/README.md (hunk 14 above)
Comment (jessemyers) [reply src/services/README.md]: vs `async`?

File: src/services/caching/__tests__/JSONStringify.test.js (hunk 12 above)
Comment (jessemyers) [reply src/services/caching/__tests__/JSONStringify.test.js]: To clarify: don't use internal service names in examples for open source software...

File: src/services/caching/index.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+
+    if (!get(config, 'cache.enabled')) {
+        if (!getMetadata().testing) {
+            terminal.disabled('caching');
+        }
+    } else {
+        if (!getMetadata().testing) {
+            terminal.enabled('caching');
+        }
+        bind('cache', () => createCacheClient(cacheHosts, get(config, 'cache.memcached')));
```
Comment (pelegdvir) [inline src/services/caching/index.js]: 1. Very memcached specific
2. `graph.cache` is quite confusing. We should use a better name here.

File: src/services/core/keys.js [hunk 17]
```diff
@@ -1,6 +1,8 @@
 import { isObject, toPairs } from 'lodash';
 import uuidv5 from 'uuid/v5';
 
+const GLOBALITY_ID = '0cffbd9f-d9bd-4dbb-88fa-327ebaec1a4b';
```
[the remark this answers is not in the document]
Comment (unorthodoxgeek) [reply src/services/core/keys.js]: oh, that's why the default cache key was different here.

**Comment by unorthodoxgeek** [comment]
closing to allow for refactoring of the code

File: src/services/README.md (hunk 14 above)
Comment (unorthodoxgeek) [reply src/services/README.md]: that's a good question, seems like the project itself is no longer on github, maybe we might want to either take over that project or replace it with something which is maintained

**Pull request closed without merging**